### ISSUE TITLE
When filtering unused schemas, schemas that are only self-referenced still show up

### ISSUE THREAD

**Issue by aneeshnazar** [thim81/openapi-format#147]
We're trying to use this filterFile:

```
{
  "methods": [],
  "inverseMethods": [],
  "tags": [],
  "inverseTags": [],
  "operationIds": [],
  "inverseOperationIds": [],
  "operations": [],
  "flags": [],
  "inverseFlags": [],
  "flagValues": [],
  "inverseFlagValues": [],
  "unusedComponents": [
    "schemas"
  ],
  "stripFlags": [],
  "responseContent": [],
  "inverseResponseContent": [],
  "requestContent": [],
  "inverseRequestContent": []
}
```

to filter our OpenAPI specs. It works for most use-cases, but schemas like this, that are not being referenced elsewhere in the filtered spec:

```
components:
  schemas:
    selfRefSchema:
      type: object
      properties:
        children:
          type: array
          items:
            $ref: "#/components/schemas/selfRefSchema"
```
are also being captured

**Comment by thim81** [comment]
hi @aneeshnazar 

Did you have some time to look at my questions?
So we can further look into this.

**Comment by shaun-jacks** [comment]
I'm also running into this issue as well. For my use-case, when removing unused schemas, I want to remove all schemas that do not show up within any of the `paths` endpoints. 

So even though a schema references itself in schemas, it is still unused within any of the `paths`

**Comment by thim81** [comment]
hi @shaun-jacks 

With "unused schemas",  you mean not used in "Paths".
The current logic checks for unused anywhere, so if a schema is still referenced in the schema section itself, for now it is still kept.

Am I correct that your use-case is to have a clean spec with only "used" schemas, that  one way or another are referenced in Paths?

**Comment by shaun-jacks** [comment]
Yeah I definitely see why with the current logic it is still kept!

Yes by unused schemas, I mean schemas not referenced at all within "Paths"! My use-case is yes to have a clean spec, where the only schemas left, are the ones referenced in Paths

**Comment by thim81** [comment]
hi @shaun-jacks 

The case explained makes total sense, so I'll try to come with a extension of the unused logic to also remove nested schemas that are not referenced in paths, and have no reason to remain in the OpenAPI document.

**Comment by shaun-jacks** [comment]
Thank you really appreciate it!

**Comment by thim81** [comment]
hi @shaun-jacks & @aneeshnazar 

I just released a new version of openapi-format (1.25.3), which has the improvements to handle unused components that are not referenced/used in any of the paths. This should result in a more compact OpenAPI, which keeps only the actual used components. 

I created a [playground](https://openapi-format-playground.vercel.app/?input=H4sIAAAAAAAAA7VXS0%2FcMBC%2B768YlUp7YsPrtDeKqrYqAgRU6tWbTFgjx05tB7o3Tv0BVS%2F9e%2FySjvO0kwDZ0uay2Xge37zHKkfJcr6Ew8XeYm%2FGZaqWMwDLrcAlXKOxcHzxib7coTZcySXsl4Q5s2vjKKMcbfkCcIO2egEwRZYxvVnCKScJTAhwZPVhzjTL0JLAhhxgF95qTJcw34lileVKorQm6iijC%2Ff6xWAyr3k0GiIz6AmZH%2Bztzbu%2FAAmaWPPclsCPQTgwKvWxuGeNLAnAuOfr7iWzeMozbsMDGEVay4g%2Blr8%2BTvfESloiCwWxPBc8Zg5cdGsIYU%2BNideYsf5Xis0mp9AwrdlmcMYtZmbIMo65UmCiq%2FK3j9k9%2BJ1lucARieTC99XhVG2NrKjm6%2FTlygwT5zhJgIHEe6eqDfm3gjLynUo2ndYxXR0hJ4WXHdvLCbQfJNC48JqPBFdvjdSoMG0ejRXDB6xqoST72zQ%2B%2F%2Fy%2FEmsHHh9%2BQcI1xlZsCFmKGmWMCXBJsXAlP%2BB5Lq%2FILXo%2B6w6cxvqsUu4IGhhVVqvVLSlv%2BoRWOWrLfd%2FwxMddMXHywQ3q9ruknjGkMlZzeTPrDD1Ttm%2Bj3JRWLkqqT%2FKOG74SSK3H8%2FFEoDuwWCxWhafCkHxl16hrJ7Sk8ZqL5KWsa5zaojpxXPMQ6YkvaSLQVKkJzqKUTbrUWG2qvg8f3l%2BTVYn70KuykrfrLP84zBUoiTSV%2FBhSAO%2FJwQiPP37WaKGQRQ9O%2BeG%2FeIlRLBklEs0YQmXAKrBuhFrNpOGW3yF9z9QdE0sP0Inj2W%2FEU4M4T18ajGHrLgUczPsiD14r8nAg8vC1Io8GIo%2B2jMSK6WEkVkoJZNILxTEYFOmulxxe1bmjS0yvgkY4UX9ZrSRyCKI%2FkkeGceWlN6NeCkC9mQ1Gg1eIVG4X51fXVRGWh%2F4oalQGY%2BNEI%2B0z5IUiplZk0kKIzXOlNF5BrZqghgJF12tugEx7fPhtgBoe1Mz9rS80xzWSzpp22Ws0lC2d6o3ZohmcnHRRx9GNx%2FtT7YkSnWhohSCwssJQ0ZTHWwHpule4VHluoMbRuqENfknkrUzDRQkMWSjQ25OovxQYNFPY394FjdLACa3asy6yIyrXKIRawr3SIpn5ixufmM3BLBnNs7NqNXQ9liUNjME%2B9NwuNLYHbbMnb1k5nUlPF0%2FPq682p9fTgpvOpMzr7jOjeN0dia5VdEkCd3WcWAFbuq7GMNlrz3aCPxKB6A3xDgAA&config=H4sIAAAAAAAAA12QMQ7CMAxFr1JlhguwVioLEgMri9W6IcKxo9hBQoi7kwaUgfG%2F%2FPwX5eVUsrnDCqS4c3fENEqMyKYdroEM8wVrzRUuikutJOGtdLjyMOwHnW8YQa%2Fsdm3wW65hBg3se5QHZoJnzx4ZMxh2EHRqtnOyIKyjEEGqxv4YE%2B8Jj797f3hstj94%2Fjo7lWKp2AnYF%2FB1wT0hUlUnsNulfYaTHHxg2OiCKxSyKSAt22nbt1zw%2FQGqHCU3OwEAAA%3D%3D) to showcase latest version.

It does contain a self-referenced schema, that is now detected and removed.

Let me know if this works for you as expected?
if it does, feel free to close this issue. If did not solve your case, provide some more details so I can reproduce your case.

x

**Comment by shaun-jacks** [comment]
Hi @thim81 thanks for working on this so quickly! I've tested this out myself.

So far, it does work as expected! However, I am noticing a performance-hit when using this for a very large OpenApi Spec. 

For example, I am using the filter multiple times for a large spec, it was taking about 10 seconds, but is now taking 4-5minutes. 

Would it be possible to add a flag to opt-out of removing self-referenced schemas, or opt-in to removing self-referenced schemas?

Depending on how large the spec is, it'll allow the user to choose if they want to use it, or if it is taking a very long time to run, they can opt-out of using the unused self-referenced schema filtering.

In summary, it does solve the issue, but the performance took a big hit for my use-cases.

**Comment by thim81** [comment]
Hi @shaun-jacks 

Thanks for sharing this vital feedback. 

This is indeed an unacceptable performance degradation. It should remain seconds, not minutes. 

Can you share how big (line numbers) the OA document is before and after the filtering?

I ll work on a new version, where I might have to introduce an option (like you suggested) or change the approach. Now I added a loop which seems to have this performance impact.

**Comment by thim81** [comment]
hi @shaun-jacks 

We just released openapi-format v 1.25.4, which should resolve the performance degradation.

Performance testing with the Stripe OpenAPI document (5MB YAML file):

1.25.2 version: 7061.96 ms or 7 sec
1.25.3 version: 298219.74 ms or 4,97min

=> 1.25.4 version: 7325.22 ms or 7,3 sec

The outcome is the same, meaning that the unused $ref items in the components should properly be removed.

before:

<img width="1788" alt="Image" src="https://github.com/user-attachments/assets/a9d069e1-21ec-465f-865d-394fb4578d9c" />

after:

<img width="1799" alt="Image" src="https://github.com/user-attachments/assets/ceedde32-87ef-4371-85f3-55fc710dc06e" />


Can you let me know, how the latest version behaves for your OpenAPI document?

**Comment by shaun-jacks** [comment]
Hi @thim81 thank you very much for the quick fix again! I can confirm the performance is now a lot better, maybe even faster than it was at 1.25.2. I am seeing both the unused circular schemas removed, and nice performance as well!

**Comment by thim81** [comment]
hi @shaun-jacks 

It sounds like openapi-format is working for you as expected, with a decent/good performance.
I'm going to close this item.

Thank you and @aneeshnazar again, for taking the time to share all your feedback and reporting the performance degredation.
Great to see that openapi-format is being used and gets better thanks to active users like yourself.
